- Description: iucv: Connection lost with high network load
  Symptom:     slow or lost IUCV-connection on systems with several CPUs
               active.
               Sometimes strange message "iucv_send returned 00000026"
               in dmesg.
  Problem:     Race in function grab_param()
  Solution:    Re-implementation of function grab_param()
  Problem-ID:  --
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/net/iucv.c	13 May 2004 12:01:39 -0000	1.27.2.2
+++ linux-2.5/drivers/s390/net/iucv.c	21 May 2004 05:21:35 -0000	1.27.2.3
@@ -1,5 +1,5 @@
 /* 
- * $Id: iucv.c,v 1.27.2.2 2004/05/13 12:01:39 braunu Exp $
+ * $Id: iucv.c,v 1.27.2.3 2004/05/21 05:21:35 braunu Exp $
  *
  * IUCV network driver
  *
@@ -29,7 +29,7 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
  *
- * RELEASE-TAG: IUCV lowlevel driver $Revision: 1.27.2.2 $
+ * RELEASE-TAG: IUCV lowlevel driver $Revision: 1.27.2.3 $
  *
  */
 
@@ -285,6 +285,7 @@
 		iparml_set_mask p_set_mask;
 	} param;
 	atomic_t in_use;
+	__u32    res;
 }  __attribute__ ((aligned(8))) iucv_param;
 #define PARAM_POOL_SIZE (PAGE_SIZE / sizeof(iucv_param))
 
@@ -351,7 +352,7 @@
 static void
 iucv_banner(void)
 {
-	char vbuf[] = "$Revision: 1.27.2.2 $";
+	char vbuf[] = "$Revision: 1.27.2.3 $";
 	char *version = vbuf;
 
 	if ((version = strchr(version, ':'))) {
@@ -469,17 +470,19 @@
 static __inline__ iucv_param *
 grab_param(void)
 {
-	iucv_param *ret;
-	static int i = 0;
+	iucv_param *ptr;
+        static int hint = 0;
 
-	while (atomic_compare_and_swap(0, 1, &iucv_param_pool[i].in_use)) {
-		i++;
-		if (i >= PARAM_POOL_SIZE)
-			i = 0;
-	}
-	ret = &iucv_param_pool[i];
-	memset(&ret->param, 0, sizeof(ret->param));
-	return ret;
+	ptr = iucv_param_pool + hint;
+	do {
+		ptr++;
+		if (ptr >= iucv_param_pool + PARAM_POOL_SIZE)
+			ptr = iucv_param_pool;
+	} while (atomic_compare_and_swap(0, 1, &ptr->in_use));
+	hint = ptr - iucv_param_pool;
+
+	memset(&ptr->param, 0, sizeof(ptr->param));
+	return ptr;
 }
 
 /**

